discord/tts.ts

```diff
@@ -308,19 +312,9 @@ export default class TTS extends EventEmitter {
 					}
 
 					await fs.writeFile(path.join(__dirname, 'tempAudio.mp3'), speech.data);
-
-					await Promise.race([
-						new Promise<void>((resolve) => {
-							const resource = createAudioResource(path.join(__dirname, 'tempAudio.mp3'));
-							this.audioPlayer.play(resource);
-							resource.playStream.on('finish', () => {
-								resolve();
-							});
-						}),
-						new Promise<void>((resolve) => {
-							setTimeout(resolve, 10 * 1000);
-						}),
-					]);
+					const resource = createAudioResource(path.join(__dirname, 'tempAudio.mp3'));
+					this.audioPlayer.play(resource);
+					await entersState(this.audioPlayer, AudioPlayerStatus.Idle, 10 * 1000).then(() => true).catch(() => false);

```

Comment:
@hakatashi こうですか？ちょっと冗長になってしまいますが

```suggestion
					const playDeferred = new Deferred<void>();
					const onFinishPlaying = () => {
						playDeferred.resolve();
					};
					await Promise.race([
						new Promise<void>(async (resolve) => {
							const resource = createAudioResource(path.join(__dirname, 'tempAudio.mp3'));
							this.audioPlayer.play(resource);
							this.audioPlayer.once(AudioPlayerStatus.Idle, onFinishPlaying);
							await playDeferred.promise;
							resolve();
						}),
						new Promise<void>((resolve) => {
							setTimeout(() => {
								this.audioPlayer.off(AudioPlayerStatus.Idle, onFinishPlaying);
								resolve();
							}, 10 * 1000);
						}),
					]);
```
```

Reply:
大枠は良さそう。いちおうPromise.raceの第一引数にはplayDeferred.promiseを直接渡すことができると思います。

Reply:
なるほど。createAudioResourceやplayなども外に出してしまってplayDeferred.promiseのみをraceに入れたほうがすっきりしますかね？

Reply:
そうですね